### ISSUE TITLE
Annotation on Service for Backend not picked up by ingress controller 

### ISSUE THREAD

**Issue by primeroz** [jcmoraisjr/haproxy-ingress#627]
When an annotation for a `backend config` is added to a `Service` it does not seem to be picked up by haproxy-ingress 0.9, when the same annotation is added to the `Ingress` it works.

In the ingress controller logs i can't see any event when the service is updated , i can see one when the ingress is updated

I am testing with `ssl-ciphers-backend` , a complete script using `kind 0.8.1` to reproduce the issue is available at https://gist.github.com/primeroz/0c3171d334da703e206f4eeb44b51158

### example
* Create Cluster and Ingress Controller and Backend Service
```
#####################################                                                                                                                               
# Backend Config 
#####################################
backend default_podinfo_https
    mode http
    balance roundrobin
    acl https-request ssl_fc
    http-request set-header X-Original-Forwarded-For %[hdr(x-forwarded-for)] if { hdr(x-forwarded-for) -m found }
    http-request del-header x-forwarded-for
    option forwardfor
    http-response set-header Strict-Transport-Security "max-age=15768000"
    server srv001 10.244.1.2:8443 weight 1 ssl verify none check inter 2s
    server srv002 127.0.0.1:1023 disabled weight 1 ssl verify none check inter 2s
    server srv003 127.0.0.1:1023 disabled weight 1 ssl verify none check inter 2s
    server srv004 127.0.0.1:1023 disabled weight 1 ssl verify none check inter 2s
    server srv005 127.0.0.1:1023 disabled weight 1 ssl verify none check inter 2s
```

* Annotate the service with `kubectl annotate service podinfo ingress.kubernetes.io/ssl-ciphers-backend=ECDHE-RSA-AES256-SHA`  ... config for backend does not change
```
#####################################
# Backend Config 
#####################################
backend default_podinfo_https
    mode http
    balance roundrobin
    acl https-request ssl_fc
    http-request set-header X-Original-Forwarded-For %[hdr(x-forwarded-for)] if { hdr(x-forwarded-for) -m found }
    http-request del-header x-forwarded-for
    option forwardfor
    http-response set-header Strict-Transport-Security "max-age=15768000"
    server srv001 10.244.1.2:8443 weight 1 ssl verify none check inter 2s
    server srv002 127.0.0.1:1023 disabled weight 1 ssl verify none check inter 2s
    server srv003 127.0.0.1:1023 disabled weight 1 ssl verify none check inter 2s
    server srv004 127.0.0.1:1023 disabled weight 1 ssl verify none check inter 2s
    server srv005 127.0.0.1:1023 disabled weight 1 ssl verify none check inter 2s
```

* Annotate the ingress with `kubectl annotate ingress podinfo ingress.kubernetes.io/ssl-ciphers-backend=ECDHE-RSA-AES256-SHA` ... config is updated
```
#####################################
# Backend Config 
#####################################
backend default_podinfo_https
    mode http
    balance roundrobin
    acl https-request ssl_fc
    http-request set-header X-Original-Forwarded-For %[hdr(x-forwarded-for)] if { hdr(x-forwarded-for) -m found }
    http-request del-header x-forwarded-for
    option forwardfor
    http-response set-header Strict-Transport-Security "max-age=15768000"
    server srv001 10.244.1.2:8443 weight 1 ssl ciphers ECDHE-RSA-AES256-SHA verify none check inter 2s
    server srv002 127.0.0.1:1023 disabled weight 1 ssl ciphers ECDHE-RSA-AES256-SHA verify none check inter 2s
    server srv003 127.0.0.1:1023 disabled weight 1 ssl ciphers ECDHE-RSA-AES256-SHA verify none check inter 2s
    server srv004 127.0.0.1:1023 disabled weight 1 ssl ciphers ECDHE-RSA-AES256-SHA verify none check inter 2s
    server srv005 127.0.0.1:1023 disabled weight 1 ssl ciphers ECDHE-RSA-AES256-SHA verify none check inter 2s
```

**Comment by primeroz** [comment]
Actually with a little more investigation it seems that 

* The `update event` on the Service is not picked up by the Ingress controller and so it does not trigger an update of the configuration
* After a time between 30s and 60s an event on the `Ingress ??` is logged and the backend is updated even though the ingress was not annotated. Seems like a _regular resync_

Is the ingress supposed to watch events on the `Service` ? 

see updated version of the sciprt to reproduce the issue: https://gist.github.com/primeroz/d36488881211738ae6f1d53ec51ad647
```
#####################################
# Annotating Service 
#####################################
CMD: kubectl annotate service podinfo ingress.kubernetes.io/ssl-ciphers-backend=ECDHE-RSA-AES256-SHA
service/podinfo annotated
#####################################
# Backend Config 
#####################################
backend default_podinfo_https
    mode http
    balance roundrobin
    acl https-request ssl_fc
    http-request set-header X-Original-Forwarded-For %[hdr(x-forwarded-for)] if { hdr(x-forwarded-for) -m found }
    http-request del-header x-forwarded-for
    option forwardfor
    http-response set-header Strict-Transport-Security "max-age=15768000"
    server srv001 10.244.1.5:8443 weight 1 ssl verify none check inter 2s
    server srv002 127.0.0.1:1023 disabled weight 1 ssl verify none check inter 2s
    server srv003 127.0.0.1:1023 disabled weight 1 ssl verify none check inter 2s
    server srv004 127.0.0.1:1023 disabled weight 1 ssl verify none check inter 2s
    server srv005 127.0.0.1:1023 disabled weight 1 ssl verify none check inter 2s


#####################################
# Backend Config - check again - sleep 60s 
#####################################
backend default_podinfo_https
    mode http
    balance roundrobin
    acl https-request ssl_fc
    http-request set-header X-Original-Forwarded-For %[hdr(x-forwarded-for)] if { hdr(x-forwarded-for) -m found }
    http-request del-header x-forwarded-for
    option forwardfor
    http-response set-header Strict-Transport-Security "max-age=15768000"
    server srv001 10.244.1.5:8443 weight 1 ssl ciphers ECDHE-RSA-AES256-SHA verify none check inter 2s
    server srv002 127.0.0.1:1023 disabled weight 1 ssl ciphers ECDHE-RSA-AES256-SHA verify none check inter 2s
    server srv003 127.0.0.1:1023 disabled weight 1 ssl ciphers ECDHE-RSA-AES256-SHA verify none check inter 2s
    server srv004 127.0.0.1:1023 disabled weight 1 ssl ciphers ECDHE-RSA-AES256-SHA verify none check inter 2s
    server srv005 127.0.0.1:1023 disabled weight 1 ssl ciphers ECDHE-RSA-AES256-SHA verify none check inter 2s
```

Logs
```

haproxy-ingress-nng76 haproxy-ingress I0723 13:13:08.975610       7 controller.go:335] Starting HAProxy update id=5
haproxy-ingress-nng76 haproxy-ingress I0723 13:13:08.986547       7 instance.go:274] HAProxy updated without needing to reload. Commands sent: 3
haproxy-ingress-nng76 haproxy-ingress I0723 13:13:08.986656       7 controller.go:386] Finish HAProxy update id=5: ingress=1.433394ms writeTmpl=7.268512ms total=8.701906ms



haproxy-ingress-nng76 haproxy-ingress I0723 13:13:46.569836       7 status.go:365] updating Ingress default/podinfo status to [{ }]
haproxy-ingress-nng76 haproxy-ingress I0723 13:13:46.575983       7 controller.go:335] Starting HAProxy update id=6
haproxy-ingress-nng76 haproxy-ingress I0723 13:13:46.576788       7 event.go:274] Event(v1.ObjectReference{Kind:"Ingress", Namespace:"default", Name:"podinfo", UID:"33c4cbb5-cce6-11ea-8fea-0242ac140002", APIVersion:"extensions/v1beta1", ResourceVersion:"751", FieldPath:""}): type: 'Normal' reason: 'UPDATE' Ingress default/podinfo
haproxy-ingress-nng76 haproxy-ingress I0723 13:13:46.625058       7 instance.go:285] HAProxy successfully reloaded
haproxy-ingress-nng76 haproxy-ingress I0723 13:13:46.625094       7 controller.go:386] Finish HAProxy update id=6: ingress=0.412665ms writeTmpl=8.367295ms reload=40.227155ms total=49.007115ms
```

**Comment by jcmoraisjr** [comment]
Hi, I can confirm that service updates doesn't start a reconciliation, so your annotation will only take effect if another update starts the process - eg scaling up or down a deployment, restart a pod, change any other annotation on any other ingress.

> Seems like a regular resync

Yup, controllers up to v0.10 has a resync every 10 minutes, this is an inheritance from the old controller which has completely dropped in v0.11.

Please confirm if:

1. forcing a reconciliation applies the service annotation - eg delete a pod from any tracked deployment to force a restart
2. you can also reproduce this issue on `v0.11-snapshot.1`

**Comment by primeroz** [comment]
Hi, on v0.11 it seems to work fine , update events on the service are caught and trigger config update ` applying 1 change notification(s): [update/service:default/podinfo]`
```
haproxy-ingress-8dg5h haproxy-ingress I0727 07:55:01.443898       7 controller.go:323] starting HAProxy update id=6
haproxy-ingress-8dg5h haproxy-ingress I0727 07:55:01.444147       7 ingress.go:188] applying 1 change notification(s): [update/service:default/podinfo]
haproxy-ingress-8dg5h haproxy-ingress I0727 07:55:01.444257       7 ingress.go:217] syncing 1 host(s) and 1 backend(s)
haproxy-ingress-8dg5h haproxy-ingress I0727 07:55:01.444795       7 instance.go:317] updating 0 host(s): []
haproxy-ingress-8dg5h haproxy-ingress I0727 07:55:01.444978       7 instance.go:334] updating 1 backend(s): [default_podinfo_https]
haproxy-ingress-8dg5h haproxy-ingress I0727 07:55:01.445261       7 dynupdate.go:139] diff outside endpoints of backend 'default_podinfo_https'
haproxy-ingress-8dg5h haproxy-ingress I0727 07:55:01.480527       7 instance.go:299] HAProxy successfully reloaded
haproxy-ingress-8dg5h haproxy-ingress I0727 07:55:01.480595       7 controller.go:355] finish HAProxy update id=6: parse_ingress=0.656227ms write_maps=0.145083ms write_config=6.637255ms reload_haproxy=29.155968ms total=36.594533ms
```


> forcing a reconciliation applies the service annotation
yeah , on 0.9 forcing a reconcile works. I deleted one of the tracked pods and the backend was properly updated with the service annotation.

## Next ? 

I am a little confused as to what i should do now :) 

I guess until either this is fixed or 0.11 is released using `service annotations` comes with this gotcha of not having an immediate update of the config , right ?

My problem is that i have multiple ingresses using the same backend ( but different secrets so can't be one ingress ) ... i wanted to avoid the risk of having the same annotation ( like `ssl-ciphers` for example ) with different values and create race conditions.

Right now i guess i should just make sure such a race condition does not exist ? 
Or maybe just apply the annotations on one of the ignresses , with the risk of loosing the config for the backend if the ingress with annotations get deleted by error

thanks!

**Comment by jcmoraisjr** [comment]
Thanks for the report and feedback!

> Next ?

Object notification is a pretty simple and straightforward fix, new v0.8, v0.9 and v0.10 versions will be released during this week. I'd suggest you to use service annotation and update the controller as soon as the new release is out.

**Comment by jcmoraisjr** [comment]
Fixed in v0.8.5, v0.9.1 and v0.10-beta.3. Closing.
